### ISSUE TITLE
SignIn Issues

### ISSUE THREAD

**Issue by MikeDev2016** [OfficeDev/skype-web-sdk-samples#9]
I'm attempting to integrate Skype for Business into my existing AngularJS web application and I am unable to sign in. My question is, what exactly are the prerequisites for developing web applications for Skype for Business Online and/or Skype Web SDK?

I've tried implementing the steps from https://msdn.microsoft.com/en-us/library/office/mt622687(v=office.16).aspx as well as this blog post that covers the same steps http://blog.thoughtstuff.co.uk/2016/03/skype-web-sdk-online-setting-everything-up-step-by-step-guide/ and yet I am faced with various lync discovery issues (see below).

The first failed request:

![image](https://cloud.githubusercontent.com/assets/19711482/15748398/be216df0-28ac-11e6-8216-405e8627b7ab.png)

This error is followed by an infinite loop that sends the following request (see below) approximately every three seconds:

![image](https://cloud.githubusercontent.com/assets/19711482/15748408/c89f17a0-28ac-11e6-82d3-f9501f1bea29.png)

However, if I leave out the part in the development document that calls signInManager.signIn with my application's Azure AD clientId and try signing in solely with the oauth2 access token I get the following errors (see below; P.S. I get the same errors if I set the domain to my Azure AD tenant, i.e. longbowadvantage).

![image](https://cloud.githubusercontent.com/assets/19711482/15748415/cf21873e-28ac-11e6-8db3-d0ecd20e1362.png)

Similarly, if I go to the online interactive Skype SDK demo at https://ucwa.skype.com/websdk# and attempt to authenticate with password, I get the following error (see below).

![image](https://cloud.githubusercontent.com/assets/19711482/15748416/d2180c10-28ac-11e6-8251-a23da8719255.png)

These errors all seem to be related to lync discover/autodiscover which leads me to believe that I am missing some prerequisites, however it is my understanding that if I am authenticating via Azure AD then Office 365 handles the Lync/Skype for Business servers and I believe my app is properly registered on Azure AD since I have already integrated Outlook (O365) into my web application which required me to register my app on Azure AD.

Any suggestions/feedback would be very much appreciated.

**Comment by sivijaya** [comment]
Mike, 
Regarding the first screenshot that you posted - the 401 is expected when you first hit that url, and your app is expected to provide an access token to authenticate. You obtain the access token when you redirect to AAD (eg. login.microsoftonline.com), authenticate and get redirected back to your app. 

The second error is one that I've seen when not using cors. This typically does not happen for a cors request so I assume it was using the xframe. For online (AAD) auth you should be using cors. Take a look at the sample that is published https://github.com/OfficeDev/skype-web-sdk-samples/tree/master/SkypeWebSDKOnlineSample, specifically these two files: 
https://github.com/OfficeDev/skype-web-sdk-samples/blob/master/SkypeWebSDKOnlineSample/index.html
https://github.com/OfficeDev/skype-web-sdk-samples/blob/master/SkypeWebSDKOnlineSample/UI/samples/scripts/sign-in.js

You'll notice a cors: true parameter to the signIn method on signInManager.

Can you try mimicking the way this sample authenticates and signs in? Also, a Fiddler trace would greatly help if you continue to hit issues.

Since you're using Azure Active Directory for authentication you should not use password based auth, rather you need to use token based auth (the flow that I explained above). Password based auth is typically used if your organization has hosted a Lync Server on their premises and has set up the necessary services (eg. lyncdiscover.yourCompany.com) to allow autodiscovery.

**Comment by MikeDev2016** [comment]
Thank you for the quick reply, I really appreciate it.

Here is a code snippet of the signIn portion of my code:

```
   var config = {
        apiKey: 'a42fcebd-5b43-4b89-a065-74450fb91255', // SDK
        apiKeyCC: '9c967f6b-a846-4df2-b43d-5167e47d81e1' // SDK+UI
    };

    Skype.initialize({ apiKey: config.apiKey }, function (api) {
        var app = new api.application;

        app.signInManager.state.changed(function (state) {
            console.log("Login State: " + state);
        });

        app.signInManager.signIn({
         "client_id": "8720fe63-a572-422c-a0af-32ede2a3a20c",  //GUID obtained from Azure app registration.
         "origins": ["https://webdir.online.lync.com/autodiscover/autodiscoverservice.svc/root"],
         "cors": true,
         "redirect_uri": 'http://localhost/defaultDashboard'  // Can be any location in the current site. (Any valid Url)
         });

        var resource = adalAuthenticationService.getResourceForEndpoint('localhost');
        var tokenStored = adalAuthenticationService.getCachedToken(resource);

        if (tokenStored) {
            var Bearercwt = 'Bearer cwt=';
            var Bearer = 'Bearer '; 
            var cwt = 'cwt';

            if (tokenStored.indexOf(cwt) == -1) {
                tokenStored = Bearercwt + tokenStored;
            }
            if (tokenStored.indexOf(Bearer) == -1) {
                tokenStored = Bearer + tokenStored;
            }

            var options = {
                auth: function (req, send) {
                    req.headers['Authorization'] = tokenStored.trim();
                    return send(req);
                },
                domain: 'longbowadvantage'
            };

            app.signInManager.signIn(options).then(
                function () {
                    console.log('Signed in as ' + app.personsAndGroupsManager.mePerson.name());
                },
                function(err) {
                    console.log('Sign in failed: '+err);
                });
        }
```

Notice how I am using adalAuthenticationService (Azure AD Authentication Library) to retrieve the token as I use the library for authentication, could this cause any issues? 

Also notice how in the first signIn request I am in fact using cors, after your suggestion I added it to my second signIn request which now attempts two extra requests to lyncdiscover/lyncdiscoverinternal, now also attempting with http:// rather than only https://. *Note: this error only occurs when the second signIn is attempted, which only happens if I comment out the first signIn attempt since the first attempt fails also.

![lyncissue](https://cloud.githubusercontent.com/assets/19711482/15759354/590fd796-28dc-11e6-8e56-f2d38e1e7818.PNG)

Also, here is a Fiddler trace (I've never used Fidler before so I'm not sure if this file actually contains the information you're interested in).

[fidlerSkype.txt](https://github.com/OfficeDev/skype-web-sdk-samples/files/296430/fidlerSkype.txt)

P.S. For temporary testing purposes I hard-coded the second signIn's "domain" field.

**Comment by faridskyman** [comment]
Having same problem, im trying to access SkypeWebSDKOnlineSample\UI, and im stuck in the loading loop

**Comment by merojo** [comment]
I'm seeing the same problem, using the onprem samples.
Using XFrame , password authentication.
Based upon other forums, it looks like the oauth request is supposed to be a "POST" , however my trace shows its an "OPTIONS" request. 
Secondly, "cors" doesn't apply to the XFrame.
Any ideas?
again, this is on premises.

**Comment by josueggh** [comment]
Same problem as @faridskyman, I don't know if you solved it ?

**Comment by merojo** [comment]
yes I did solve this, basically its a setup issue.
the lync server's certificate needs to be installed on the web browser.  But that certificate is issued by the web server.  So the lync/skype server has to know the URL of the certificate authority ( the web server ) , and add that as a trusted domain, if memory serves correctly, HTH.
Then when the browser first opens the lync/skype URL , it should prompt to install a certificated.  Or you can install the certificate manually.   How this is going to work in production , I dunno.  THe config of skype is more of a sys admin job than a developer job.

**Comment by sivijaya** [comment]
@MikeDev2016 
1. Why are you trying to sign in twice in the same sample? 
2. Your first sign in approach (using client_id, cors, origins) indicates that your SFB account is hosted Online. However your second signin snippet using the domain parameter indicates an onprem approach. You need to choose either one. If it is an online account then you must use the first approach, if it is an onprem account then you can use username/password auth.

In Fiddler you need to enable https decryption. Without this your trace will not capture any https requests.
Please don't attach your Fiddler trace here as it will likely contain sensitive information with https decryption turned on.

Does the adalAuthenticationService do a full page redirect in order to sign in to Azure AD online? 
Basically, your sign in flow should look like this (for a user who is hosted on SFB online):
1. user navigates to your web app. You do a full page redirect to Azure AD (login.microsoftonline.com) for the user to sign in.
2. AAD redirects back to your web app with access token in the location hash. This can be ignored.
3. You call Skype's sign in method with the correct parameters (client_id, cors, redirect_uri, origins).
4. You monitor signInManager's state to check whether you are successfully signed in.

**Comment by sivijaya** [comment]
@faridskyman , @josueggh 
Are you trying to sign in to SFB online or onprem?
Typically the loading loop occurs when SFB encounters an error in the response from Azure AD when SFB tries to sign in. To diagnose this look at your fiddler trace and look for a 200 OK response from Azure AD. Look at the very end of the response for any particular diag codes.

I've done a walkthrough of building a sample that signs in to SFB online at Microsoft Ignite this year, you can [watch the video](https://techcommunity.microsoft.com/t5/Microsoft-Ignite-Content/BRK4012-Dive-Deep-on-Skpe-Web-SDK-amp-Skype-for-Business-App-SDK/td-p/10393). Specifically, look around 18minutes where I show a similar error from Azure AD.

**Comment by KthProg** [comment]
@merojo Can you explain in more detail how you fixed the issue? I'm getting the same kinds of errors.

**Comment by manoj1kumar1** [comment]
Hello Sivijaya

I am using SkypeWebSDKOnlineSample and receiving #access_token as well but Skype initialization producing error as following :

![image](https://user-images.githubusercontent.com/2906791/27003454-60cb81b6-4e14-11e7-9e97-88af3fbd63a4.png)

and stuck on sign-in step. 

I don't understand where I am doing wrong even there is not showing any error. 

Your help will be appreciated.

**Comment by KthProg** [comment]
**My advice to everyone:** just use the C# SDK if you want to create a bot. It's way easier. Their web SDK is painfully complicated and poorly documented.

**Comment by manoj1kumar1** [comment]
@kthprog  thanks for your comment . Is it possible to use  c# SDK with php based web application ..?

**Comment by KthProg** [comment]
@manoj1kumar1 Afraid not. There are Node.js and REST API versions though. I haven't tried these however, so I can't say whether they work as well as the C# SDK. For your purposes, you'd probably want to look at making HTTP requests via PHP using their REST SDK. [Here is the link](https://docs.microsoft.com/en-us/bot-framework/dotnet/bot-builder-dotnet-overview). (You can change the SDK type at the top)

**Comment by shadab280296** [comment]
implemented every thing is working fine i have some issue regarding skype audio call 
it get ended when i make an audio call with SIP that i have set 
![screenshot 1](https://user-images.githubusercontent.com/16804412/31541119-43d52ba6-b02b-11e7-82a7-ad230160f861.png)
please any suggestion would be appreciated

**Comment by tntstn** [comment]
Hi everyone,

I have another problem about SigninManager. when I login in like this;

tan.tastan@abcd.com

abdc.com is reachable. but If I write wrong domain for example; tan.tastan@abcdE.com I am not getting a response and my application is waiting. nothing happens. here is my sample code;

setting include username, password, domain information.

function doLogin(settings) {
    return new Promise((resolve, reject) => {
        window.skypeWebSDKApi.signInManager.signIn(settings).then((response) => {
            resolve(response);
        }, (error) => {
            reject(error);
        }).catch(reject);
    });
}

what is the problem? what do you think?

**Comment by manoj1kumar1** [comment]
@KthProg : have you get fixed your audio call issues? I am looking for programmer who can implement skype for business online with our php based web-application, I am ready to pay accordingly. 
please message me directly at this email : manoj.techlect@gmail.com
Thanks

**Comment by atanevski** [comment]
Hello, I have sign in issues, and i am stuck in SigningIn state 

These are the scenarios that I have tried so far

1. User authenticates with Azure AAD to my web application and he is logged in successfully on registered app and another registered app that has Skype permissions and i am getting this error 
![image](https://user-images.githubusercontent.com/15204761/33199851-77e36010-d0f2-11e7-8f40-c627d413e394.png)
using this code 
![image](https://user-images.githubusercontent.com/15204761/33200963-fbb734f8-d0f6-11e7-800c-91499b848a02.png)
and this is the error 
![image](https://user-images.githubusercontent.com/15204761/33201003-1b2c09d0-d0f7-11e7-9946-1c4ecc996ca7.png)
2. Same as 1 but i am using one registered app that has both AAD and Skype required permissions 
![image](https://user-images.githubusercontent.com/15204761/33201003-1b2c09d0-d0f7-11e7-9946-1c4ecc996ca7.png)
3. I acquired the token manually and tried to use WebTicket
the console does not have errors, some JMigrations console logs but still stuck in SiginingIn state

**Comment by pmironov** [comment]
@atanevski  hi, try to add http://<your_app_url>/empty.html to replyUrls section in your app's manifest in Azure. That is what helped me.

**Comment by atanevski** [comment]
@pmironov no luck
![image](https://user-images.githubusercontent.com/15204761/33213454-f127d7aa-d127-11e7-89fe-4f2fa728ec73.png)
![image](https://user-images.githubusercontent.com/15204761/33213473-022c55da-d128-11e7-8854-39939bc32803.png)

Just to be clear ... So i am currently logged in with Azure AD on my web portal and i am just executing this Skype peace of code and i should be logged in. Am i doing it right or am i missing something?

Thanks

**Comment by pmironov** [comment]
@atanevski I meant the app manifest in Azure where you set replyUrls not the config of samples app

**Comment by atanevski** [comment]
@pmironov  Thank you

**Comment by jarvid488** [comment]
Hi , 
I am signing with anonymous user in skype for business i have followed this link 
i am successful to login with my licensed account and generate the meeting uri 
What if followed from the blog that to sigin anonymously we have to sigin from Fromm meeting uri 


I am getting an Error Command disabled can anyone help
